### PR TITLE
feat(parser): Support for API GW v1 proxy schema & envelope

### PR DESCRIPTION
**Pull request by risenberg-cyberark**
This schema exists in the validator but not in the parser. It was requested in the webinar audience.
This feature will also provide an envelope class which will parse the string body parameter as a pydantic Model.

Also added new identity parameters as defined at https://docs.aws.amazon.com/apigateway/latest/developerguide/http-api-develop-integrations-lambda.html

### CODE REVIEW

File: aws_lambda_powertools/utilities/parser/envelopes/apigw.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+from ..models import APIGatewayProxyEventModel
+from ..types import Model
+from .base import BaseEnvelope
+
+logger = logging.getLogger(__name__)
+
+
+class ApiGatewayEnvelope(BaseEnvelope):
+    """Api Gateway envelope to extract data within detail key"""
```
Comment (heitorlessa) [inline aws_lambda_powertools/utilities/parser/envelopes/apigw.py:12]:
```suggestion
    """API Gateway envelope to extract data within body key"""
```

File: aws_lambda_powertools/utilities/parser/models/apigw.py [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+    @root_validator()
+    def check_message_id(cls, values):
+        message_id, event_type = values.get("messageId"), values.get("eventType")
+        if message_id is not None and event_type != "MESSAGE":
+            raise TypeError("messageId is available only when the `eventType` is `MESSAGE`")
+        return values
+
+    @root_validator(pre=True)
+    def check_both_http_methods(cls, values):
```
Comment (heitorlessa) [inline aws_lambda_powertools/utilities/parser/models/apigw.py:96]: could you add a note why this check is needed?

File: aws_lambda_powertools/utilities/parser/models/apigw.py [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+        http_method, req_ctx_http_method = values.get("httpMethod"), values.get("requestContext", {}).get(
+            "httpMethod", ""
+        )
+        if http_method != req_ctx_http_method:
+            raise TypeError("httpMethods and requestContext.httpMethod must be equal")
+        return values
+
+    @root_validator(pre=True)
+    def check_both_paths(cls, values):
```
Comment (heitorlessa) [inline aws_lambda_powertools/utilities/parser/models/apigw.py:105]: could you add a note why this check is needed?

File: aws_lambda_powertools/utilities/parser/models/apigw.py [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+    queryStringParameters: Optional[Dict[str, str]]
+    multiValueQueryStringParameters: Optional[Dict[str, List[str]]]
+    requestContext: APIGatewayEventRequestContext
+    pathParameters: Optional[Dict[str, str]]
+    stageVariables: Optional[Dict[str, str]]
+    isBase64Encoded: bool
+    body: str
+
+    @root_validator()
+    def check_message_id(cls, values):
```
Comment (heitorlessa) [inline aws_lambda_powertools/utilities/parser/models/apigw.py:89]: could you add a note why this check is needed? e.g. spoofing?

File: docs/utilities/parser.md [hunk 5]
```diff
@@ -161,6 +161,7 @@ Parser comes with the following built-in models:
 | **KinesisDataStreamModel** | Lambda Event Source payload for Amazon Kinesis Data Streams        |
 | **SesModel**               | Lambda Event Source payload for Amazon Simple Email Service        |
 | **SnsModel**               | Lambda Event Source payload for Amazon Simple Notification Service |
+| **APIGatewayProxyEvent**   | Lambda Event Source payload for Amazon API Gateway                 |
```
Comment (heitorlessa) [inline docs/utilities/parser.md:164]: API Gateway v1, v2, or both? e.g. REST vs HTTP API. By my first look I'd guess REST API

File: docs/utilities/parser.md [hunk 6]
```diff
@@ -294,16 +295,16 @@ Here's an example of parsing a model found in an event coming from EventBridge,
...
+| **SnsEnvelope**                                                                                                                              | 1. Parses data using `SnsModel`. <br/> 2. Parses records in `body` key using your model and return them in a list.                                                                                          | `List[Model]`                      |
+| **SnsSqsEnvelope**                                                                                                                           | 1. Parses data using `SqsModel`. <br/> 2. Parses SNS records in `body` key using `SnsNotificationModel`. <br/> 3. Parses data in `Message` key using your model and return them in a list.                  | `List[Model]`                      |
+| **ApiGatewayEnvelope**         1. Parses data using `APIGatewayProxyEventModel`. <br/> 2. Parses `body` key using your model and returns it. | `Model`                                                                                                                                                                                                     |
```
Comment (heitorlessa) [inline docs/utilities/parser.md:307]: API Gateway v1, v2, or both? e.g. REST vs HTTP API. By my first look I'd guess REST API

**Review by heitorlessa — CHANGES_REQUESTED** [review]
Awesome, thank you @risenberg-cyberark. A few minor questions on the additional checks, and an explicit comment to tag whether this is REST API (v1) or HTTP API (v2), or both, so it's easier to add and use HTTP API later judging by the import, naming, etc.

File: aws_lambda_powertools/utilities/parser/models/apigw.py (hunk 2 above)
Comment (risenberg-cyberark) [reply aws_lambda_powertools/utilities/parser/models/apigw.py:96]: what would you like to write here? basically what i saw that in the v1 version there's a duplication 3 fields, so i made sure that all values are actually the same.

File: aws_lambda_powertools/utilities/parser/models/apigw.py (hunk 3 above)
Comment (risenberg-cyberark) [reply aws_lambda_powertools/utilities/parser/models/apigw.py:105]: what would you like to write here? basically what i saw that in the v1 version there's a duplication 3 fields, so i made sure that all values are actually the same. V2 removes this duplication

File: aws_lambda_powertools/utilities/parser/models/apigw.py (hunk 4 above)
Comment (risenberg-cyberark) [reply aws_lambda_powertools/utilities/parser/models/apigw.py:89]: what would you like to write here? basically what i saw that in the v1 version there's a duplication 3 fields, so i made sure that all values are actually the same. V2 removes this duplication

File: docs/utilities/parser.md (hunk 5 above)
Comment (risenberg-cyberark) [reply docs/utilities/parser.md:164]: V1

**Comment by heitorlessa** [comment]
Thanks for the clarifications @risenberg-cyberark - I'm happy to merge after we remove those root validators.

I don't see a need a benefit in keeping them if we, the provider, are the ones duplicating these possibly for backward compatible reasons (until HTTP API when they're gone).

Alternatively, I can merge and fiz that later too - up to you.

Thanks a lot for adding this support!

**Comment by risenberg-cyberark** [comment]
> Thanks for the clarifications @risenberg-cyberark - I'm happy to merge after we remove those root validators.
> 
> I don't see a need a benefit in keeping them if we, the provider, are the ones duplicating these possibly for backward compatible reasons (until HTTP API when they're gone).
> 
> Alternatively, I can merge and fiz that later too - up to you.
> 
> Thanks a lot for adding this support!

I'll explain. They are duplicated in the payload. I didnt duplicate them, i just made sure that the values are the same since they clearly are supposed to be the same. Those 2 items reside on the main event and are also found in the request context part. 

You can keep them or remove them but I do think they add a value becuase if there's a discrepancy in between them, you should not use the entire payload at all ,

**Comment by heitorlessa** [comment]
> > Thanks for the clarifications @risenberg-cyberark - I'm happy to merge after we remove those root validators.
> 
> > 
> 
> > I don't see a need a benefit in keeping them if we, the provider, are the ones duplicating these possibly for backward compatible reasons (until HTTP API when they're gone).
> 
> > 
> 
> > Alternatively, I can merge and fiz that later too - up to you.
> 
> > 
> 
> > Thanks a lot for adding this support!
> 
> 
> 
> @heitorlessa 
> 
> I'll explain. They are duplicated in the payload. I didnt duplicate them, i just made sure that the values are the same since they clearly are supposed to be the same. Those 2 items reside on the main event and are also found in the request context part. 
> 
> 
> 
> You remove them but I do think they add a value because if there's a discrepancy in between them, you should not use the entire payload at all ,

That's what I understood too @risenberg-cyberark ;)

Putting in another way, I wouldn't like to break customers applications because this behaviour isn't documented. For this reason, I don't trust that API Gateway Lambda integration to follow through these checks consistently.

**Comment by risenberg-cyberark** [comment]
> > > Thanks for the clarifications @risenberg-cyberark - I'm happy to merge after we remove those root validators.
> > 
> > 
> > > 
> > 
> > 
> > > I don't see a need a benefit in keeping them if we, the provider, are the ones duplicating these possibly for backward compatible reasons (until HTTP API when they're gone).
> > 
> > 
> > > 
> > 
> > 
> > > Alternatively, I can merge and fiz that later too - up to you.
> > 
> > 
> > > 
> > 
> > 
> > > Thanks a lot for adding this support!
> > 
> > 
> > @heitorlessa
> > I'll explain. They are duplicated in the payload. I didnt duplicate them, i just made sure that the values are the same since they clearly are supposed to be the same. Those 2 items reside on the main event and are also found in the request context part.
> > You remove them but I do think they add a value because if there's a discrepancy in between them, you should not use the entire payload at all ,
> 
> That's what I understood too @risenberg-cyberark ;)
> 
> Putting in another way, I wouldn't like to break customers applications because this behaviour isn't documented. For this reason, I don't trust that API Gateway Lambda integration to follow through these checks consistently.

np, removed the relevant root validator. is the last root validator ok?

File: aws_lambda_powertools/utilities/parser/models/apigw.py [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+    resourcePath: str
+    domainName: Optional[str]
+    domainPrefix: Optional[str]
+    extendedRequestId: Optional[str]
+    httpMethod: Literal["DELETE", "GET", "HEAD", "OPTIONS", "PATCH", "POST", "PUT"]
+    path: str
+    connectedAt: Optional[datetime]
+    connectionId: Optional[str]
+    eventType: Optional[Literal["CONNECT", "MESSAGE", "DISCONNECT"]]
+    eventType: Optional[str]
```
Comment (heitorlessa) [inline aws_lambda_powertools/utilities/parser/models/apigw.py:66]: Is this perhaps a typo as it's defined twice?

That's for WebSockets and the line 65 is correct

**Review by heitorlessa — CHANGES_REQUESTED** [review]
All good on the validators except I think there's an accidental duplicates field (eventType).

Mind checking before I merge?

File: aws_lambda_powertools/utilities/parser/models/apigw.py (hunk 7 above)
Comment (risenberg-cyberark) [reply aws_lambda_powertools/utilities/parser/models/apigw.py]: seems like a bad copy paste :)
removed

**Comment by heitorlessa** [comment]
I bet :D Glad I checked w/ you before merging :) Merging now!

**Review by heitorlessa — APPROVED** [review]

**Pull request merged**